### ISSUE TITLE
[QUESTION] Timeout exception on get_file(file_id) after long delay between bot interactions

### ISSUE THREAD

**Issue by awerks** [python-telegram-bot/python-telegram-bot#3830]
### Issue I am facing

Hello,

I've encountered an issue when using get_file(file_id) method in the python-telegram-bot library. Whenever there is a long delay between bot interactions, I get a timeout error. However, on the second attempt, it works normally.

I'm running the bot in polling mode with local mode turned on. Here are the logs of the error

### Traceback to the issue

```python
INFO - HKS (id: #######) has initalized a conversation.
﻿INFO - Downloading a file...
﻿Traceback (most recent call last):
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_backends/anyio.py", line 34, in read
﻿return await self._stream.receive(max_bytes=max_bytes)
﻿File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 1203, in receive
﻿await self._protocol.read_event.wait()
﻿File "/usr/lib/python3.10/asyncio/locks.py", line 214, in wait
﻿await futThe above exception was the direct cause of the following exception:
﻿Traceback (most recent call last):
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/request/_httpxrequest.py", line 216, in do_request
﻿res = await self._client.request(
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_client.py", line 1530, in request
﻿return await self.send(request, auth=auth, follow_redirects=follow_redirects)
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_client.py", line 1617, in send
﻿response = await self._send_handling_auth(
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_client.py", line 1645, in _send_handling_auth
﻿response = await self._send_handling_redirects(
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_client.py", line 1682, in _send_handling_redirects
﻿response = await self._send_single_request(request)
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_client.py", line 1719, in _send_single_request
﻿response = await transport.handle_async_request(request)
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_transports/default.py", line 352, in handle_async_request
﻿with map_httpcore_exceptions():
﻿File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
﻿self.gen.throw(typ, value, traceback)
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
﻿raise mapped_exc(message) from exc
﻿httpx.ReadTimeout
﻿The above exception was the direct cause of the following exception:
﻿Traceback (most recent call last):
﻿File "/app/bot.py", line 866, in handle_video_or_document
﻿file = await context.bot.get_file(file_id)
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/ext/_extbot.py", line 1654, in get_file
﻿return await super().get_file(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/_bot.py", line 388, in decorator
﻿result = await func(self, *args, **kwargs)  # skipcq: PYL-E1102
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/_bot.py", line 3047, in get_file
﻿result = await self._post(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/_bot.py", line 476, in _post
﻿return await self._do_post(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/ext/_extbot.py", line 332, in _do_post
﻿return await super()._do_post(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/_bot.py", line 504, in _do_post
﻿return await request.post(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/request/_baserequest.py", line 168, in post
﻿result = await self._request_wrapper(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/request/_baserequest.py", line 288, in _request_wrapper
﻿raise exc
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/request/_baserequest.py", line 278, in _request_wrapper
﻿code, payload = await self.do_request(
﻿File "/usr/local/lib/python3.10/dist-packages/telegram/request/_httpxrequest.py", line 233, in do_request
﻿raise TimedOut from err
﻿telegram.error.TimedOut: Timed out
﻿INFO - HKS (id: 1588840053) has initalized a conversation.
﻿INFO - Downloading a file...
﻿INFO - Downloaaded succesfully...
﻿asyncio.exceptions.CancelledError
﻿During handling of the above exception, another exception occurred:
﻿Traceback (most recent call last):
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_exceptions.py", line 10, in map_exceptions
﻿yield
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_backends/anyio.py", line 32, in read
﻿with anyio.fail_after(timeout):
﻿File "/usr/local/lib/python3.10/dist-packages/anyio/_core/_tasks.py", line 119, in __exit__
﻿raise TimeoutError
﻿TimeoutError
﻿The above exception was the direct cause of the following exception:
﻿Traceback (most recent call last):
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
﻿yield
﻿File "/usr/local/lib/python3.10/dist-packages/httpx/_transports/default.py", line 353, in handle_async_request
﻿resp = await self._pool.handle_async_request(req)
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/connection_pool.py", line 262, in handle_async_request
﻿raise exc
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/connection_pool.py", line 245, in handle_async_request
﻿response = await connection.handle_async_request(request)
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/connection.py", line 96, in handle_async_request
﻿return await self._connection.handle_async_request(request)
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/http11.py", line 121, in handle_async_request
﻿raise exc
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/http11.py", line 99, in handle_async_request
﻿) = await self._receive_response_headers(**kwargs)
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/http11.py", line 164, in _receive_response_headers
﻿event = await self._receive_event(timeout=timeout)
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_async/http11.py", line 200, in _receive_event
﻿data = await self._network_stream.read(
﻿File "/usr/local/lib/python3.10/dist-packages/httpcore/_exceptions.py", line 14, in map_exceptions
﻿raise to_exc(exc) from exc
﻿httpcore.ReadTimeout
INFO - HKS (id: #########) has initalized a conversation.
﻿INFO - Downloading a file...
﻿INFO - Downloaaded succesfully...
```

### Related part of your code

```python
logger.info(f"{context.user_data['name']} (id: {context.user_data['user_id']}) has initalized a conversation.")
        logger.info("Downloading a file...")
        context.user_data["message"] = await update.message.reply_text(get(context, "downloading_video_text"))
        if update.message.video:
            # The message is a video
            file_id = update.message.video.file_id
        elif update.message.document:
            # The message is a document
            file_id = update.message.document.file_id

        file = await context.bot.get_file(file_id)
        logger.info("Downloaded succesfully...")



###########
    application = Application.builder().token(TOKEN).base_url("http://127.0.0.1:8081/bot").base_file_url("http://127.0.0.1:8081/file/bot").local_mode(True).build()

application.run_polling(drop_pending_updates = True)
```

### Operating System

Linux Ubuntu

### Version of Python, python-telegram-bot & dependencies

```shell
Bot API 6.7
Python 3.11.4
```

**Comment by Bibo-Joshi** [comment]
Hi. PTB does not actively close connections after some idle time or anything like that though I'm not completely familiar on whether the `httpx` library does something like this. Are you hosting your bot on a local machine/a server that you control yourself or are you hosting it in some rented server/vps/cloud service?

**Comment by awerks** [comment]
@Bibo-Joshi it is hosted on a cloud service. It gets the update, but times out the first time for some reason.

**Comment by Bibo-Joshi** [comment]
Could you try if you can reproduce the issue in a local machine? I recall that e.g. heroin has/had a tendency to put inactive processes into some idling mode, so I could imagine that it has something to do with your provider.

**Comment by kho239** [comment]
I have the same issue. Steps to reproduce:
1. Start a local bot api server
2. Run this script with your TOKEN:
`from telegram import Update
from telegram.ext import ApplicationBuilder, ContextTypes, MessageHandler, \
    filters

application = ApplicationBuilder()\
              .token(TOKEN)\
              .base_url("http://127.0.0.1:8081/bot")\
              .local_mode(True)\
              .base_file_url("http://127.0.0.1:8081/file/bot").build()


async def handle_file(update: Update, context: ContextTypes.DEFAULT_TYPE):
    file_id = update.message.document.file_id
    new_file = await context.bot.get_file(file_id)
    local_file_path = await new_file.download_to_drive()
    await update.message.reply_text(str(local_file_path))


application.add_handler(MessageHandler(filters.Document.ZIP, handle_file))
application.run_polling()`
3. Send a big .zip file to your bot (for me worked ~80 Mb)
4. This error pops up:
`No error handlers are registered, logging exception.
Traceback (most recent call last):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 34, in read
    return await self._stream.receive(max_bytes=max_bytes)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 1203, in receive
    await self._protocol.read_event.wait()
  File "/opt/homebrew/Cellar/python@3.10/3.10.12_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/locks.py", line 214, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 32, in read
    with anyio.fail_after(timeout):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/anyio/_core/_tasks.py", line 119, in __exit__
    raise TimeoutError
TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
    yield
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 262, in handle_async_request
    raise exc
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 245, in handle_async_request
    response = await connection.handle_async_request(request)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/connection.py", line 96, in handle_async_request
    return await self._connection.handle_async_request(request)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/http11.py", line 121, in handle_async_request
    raise exc
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/http11.py", line 99, in handle_async_request
    ) = await self._receive_response_headers(**kwargs)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/http11.py", line 164, in _receive_response_headers
    event = await self._receive_event(timeout=timeout)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_async/http11.py", line 200, in _receive_event
    data = await self._network_stream.read(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_backends/anyio.py", line 31, in read
    with map_exceptions(exc_map):
  File "/opt/homebrew/Cellar/python@3.10/3.10.12_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ReadTimeout

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/request/_httpxrequest.py", line 216, in do_request
    res = await self._client.request(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_client.py", line 1530, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_client.py", line 1617, in send
    response = await self._send_handling_auth(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_client.py", line 1645, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_client.py", line 1682, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_client.py", line 1719, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_transports/default.py", line 352, in handle_async_request
    with map_httpcore_exceptions():
  File "/opt/homebrew/Cellar/python@3.10/3.10.12_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/ext/_application.py", line 1173, in process_update
    await coroutine
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/ext/_basehandler.py", line 141, in handle_update
    return await self.callback(update, context)
  File "/Users/kho239/PycharmProjects/PTBfileBug/main.py", line 16, in handle_file
    new_file = await context.bot.get_file(file_id)
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/ext/_extbot.py", line 1654, in get_file
    return await super().get_file(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/_bot.py", line 388, in decorator
    result = await func(self, *args, **kwargs)  # skipcq: PYL-E1102
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/_bot.py", line 3047, in get_file
    result = await self._post(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/_bot.py", line 476, in _post
    return await self._do_post(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/ext/_extbot.py", line 332, in _do_post
    return await super()._do_post(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/_bot.py", line 504, in _do_post
    return await request.post(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/request/_baserequest.py", line 168, in post
    result = await self._request_wrapper(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/request/_baserequest.py", line 288, in _request_wrapper
    raise exc
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/request/_baserequest.py", line 278, in _request_wrapper
    code, payload = await self.do_request(
  File "/Users/kho239/PycharmProjects/PTBfileBug/yar-python-client/venv/lib/python3.10/site-packages/telegram/request/_httpxrequest.py", line 233, in do_request
    raise TimedOut from err
telegram.error.TimedOut: Timed out`

**Comment by Poolitzer** [comment]
Ah hi, that is not an issue, that is everything working as expected :)

The reason behind your error is because you request large files. The local bot api gets this request and only answers it when the file is fully downloaded (obviously, you expect a non broken file after a non errored request response). Depending on the download speed, this can take a while with bigger files.

PTB has a timeout time to stop waiting until a response occurs. This is there to let your code finish/wait for the next update when the connection to telegram suddenly gets interrupted for example, and to let the error handlers do their thing if you registered them.

So since you request a large enough file so the bot api instance takes a hot minute to respond to the request, PTB errors out. Everything works as expected.

You should overwrite this time to an expected one on a case by case basis using [get_file(read_timeout)](https://docs.python-telegram-bot.org/en/stable/telegram.bot.html#telegram.Bot.get_file.params.read_timeout). Don't just set it to a super high one because in the event the connection actually gets dropped your bot will wait till then (also dont set it to False I think, where it just waits forever, since then the bot just hangs). Calculate it by the file size and your average download speed of telegram files.

**Comment by kho239** [comment]
> Ah hi, that is not an issue, that is everything working as expected :)

Well, probably you could make the raised error more verbose, it was quite hard to understand what's going on based on the error `telegram.error.TimedOut: Timed out` :)
Or even maybe increase the default wait time in case a local mode is on (guess I'm not the only one who transferred to a local bot api for that reason)

**Comment by Bibo-Joshi** [comment]
> > Ah hi, that is not an issue, that is everything working as expected :)
> 
> Well, probably you could make the raised error more verbose, it was quite hard to understand what's going on based on the error `telegram.error.TimedOut: Timed out` :)

PTB pretty much has the same information available as you have via the traceback, so there is not really and additional information that we can provide. What pool explained is general timeouts that can happen in any networking operation. They are surely more likely to occur in operations with large files but nevertheless can also occur in any other operation. We have a dedicated wiki page on handling network errors.please have a look at that.

> Or even maybe increase the default wait time in case the local mode is on (guess I'm not the only one who transferred to a local bot api because of big files)
> 

As pool explained, too large timeouts are not good for operations that don't need them. Sending larger files is one use case of local bot API severs, but not the only one. I wouldn't like to change the default value in PTB itself. We do however allow for the user (you) to set a custom default timeout for their own bot. Please have a look at the wiki page on adding defaults.

I want to add that your issue is in fact not the same as the original authors as your timeout apparently happens every time and not only after a period of inactivity.

**Comment by Poolitzer** [comment]
Please actually open a new issue for that now :)

**Comment by Poolitzer** [comment]
@kho239 https://t.me/pythontelegrambotgroup/696401

btw, why do we lock old issues? I dont like it being done by default, its against githubs idea of that feature as well from my understanding, https://docs.github.com/en/communities/moderating-comments-and-conversations/locking-conversations
